### PR TITLE
feat: Add reactive toggle switch for resource active state (TREK-134)

### PR DESCRIPTION
**Pull request by aro-ssystems**
## Summary

Implements reactive toggle switch for resource active/inactive state in resources table.

## Changes

- **Template**: Replaced static badges with Bootstrap 4 custom-switch toggle in `resource_item_row.mustache`
- **JavaScript**: Added `_handleToggleActive()` method in `resource_catalog.js`
- **Event Handler**: Added 'change' event delegation for `[data-action='toggle-active']` in table view

## Implementation

Submits resource updates via direct Ajax call to `core_form_dynamic_form`:
1. Toggle change triggers `_handleToggleActive()`
2. Builds form data object with current resource data + toggled active value
3. Submits via Ajax.call() to `edit_resource_form`
4. Response updates reactive state automatically via `stateManager.processUpdates()`
5. UI updates instantly without page reload

## Testing

- Toggle ON→OFF: Checkbox unchecks, opacity-50 applied, DB updated
- Toggle OFF→ON: Checkbox checks, opacity removed, DB updated
- Label clicks work correctly (triggers 'change' event)
- State persists across page reloads
- Reactive state updates table immediately

## Technical Notes

- **Bootstrap Version**: Uses Bootstrap 4 `custom-control custom-switch` markup (correct for Moodle 4.5)
- **Event Handling**: Uses 'change' event instead of 'click' to properly handle label clicks
- **Async/Await**: All async operations properly awaited to prevent race conditions

## Related

- TREK-134: Implement reactive toggle switch for resource active state
- Follows reactive state management pattern
- EPIC-16, EPIC-18: Code review fixes applied

### CODE REVIEW

**Review by aro-ssystems — COMMENTED** [review]
## Code Review Fixes Complete ✅

All 9 review comments have been addressed with 7 commits:

### Critical Fixes (Priority 0)
- **0b8037b**: Fixed roomids data loss - Now parsing JSON from data-item-roomids and serializing as repeated parameters (roomids[])
- **124cdcb**: Fixed data-itemid selector mismatch - Updated to match template attribute (no hyphen)
- **7d3a54b**: Fixed async race condition - Made _handleItemCategoryChanged async with await
- **82fd152**: Fixed preventDefault desync - Removed to allow natural checkbox toggle

### Quality Improvements (Priority 1)
- **85050e7**: Fixed i18n - Cache localized active/inactive strings from core
- **8728643**: Added null checks - Prevents crashes in _replaceRenderedItem
- **031fea4**: Fixed category rendering - Now awaits _render() for new categories

### No Changes Required
- **Bootstrap version**: Confirmed Moodle 4.5 uses Bootstrap 4.6.2 - current markup is correct
- **roomids/description init**: Already resolved by roomids parsing fix

I'll respond to individual comments with commit references.

**Review by aro-ssystems — COMMENTED** [review]
All 9 review comments have been addressed. Please review the changes:

- 7 commits with code fixes
- 2 issues required no changes (Bootstrap version correct, roomids already fixed)

Ready for re-review. @copilot-pull-request-reviewer

File: amd/src/resource_category.js [hunk 1]
```diff
@@ -141,25 +175,19 @@ export default class ResourceCategory extends BaseComponent {
...
         if (hadItem !== shouldHaveItem) {
             // Item moved - re-render.
             this._renderItems();
-        } else if (shouldHaveItem) {
-            // Item updated in place.
-            const component = this.itemComponents.get(element.id);
-            if (component) {
-                component.update(element);
-            }
         }
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply amd/src/resource_category.js:191]: ✅ **Fixed in commit 7d3a54b**

Made `_handleItemCategoryChanged` async and added `await this._renderItems()` to prevent race conditions and unhandled promise rejections.

File: amd/src/resource_category.js [hunk 2]
```diff
@@ -118,13 +119,46 @@ export default class ResourceCategory extends BaseComponent {
...
+    _initItemsFromDOM() {
+        if (!this.categoryElement) {
+            return;
+        }
+
+        const state = this.reactive.state;
+        const itemRows = this.categoryElement.querySelectorAll('tr[data-item-id]');
+
+        itemRows.forEach(itemRow => {
+            const itemId = parseInt(itemRow.dataset.itemId);
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply amd/src/resource_category.js]: ✅ **Fixed in commit 124cdcb**

Updated selector to `tr[data-itemid]` and dataset access to `itemid` (no camelCase) to match template attribute.

File: amd/src/resource_catalog.js [hunk 3]
```diff
@@ -270,6 +280,12 @@ export default class extends BaseComponent {
         tableView.addEventListener('click', async(e) => {
             const target = e.target.closest('button[data-action]');
             if (!target) {
+                // Check for toggle switches.
+                const toggle = e.target.closest('[data-action="toggle-active"]');
+                if (toggle) {
+                    e.preventDefault();
+                    await this._handleToggleActive(toggle);
+                }
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply amd/src/resource_catalog.js]: ✅ **Fixed in commit 82fd152**

Removed `preventDefault()` call to allow natural checkbox toggle behavior.

File: amd/src/resource_catalog.js [hunk 4]
```diff
@@ -455,6 +471,60 @@ export default class extends BaseComponent {
...
+            amountirrelevant: item.amountirrelevant ? 1 : 0,
+            sortorder: item.sortorder,
+            active: newActiveState ? 1 : 0,
+            roomids: item.roomids || [],
+            action: 'put',
+            [`_qf__mod_bookit_form_edit_resource_form`]: 1
+        };
+
+        // Convert to URL-encoded string.
+        const formDataString = new URLSearchParams(formData).toString();
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply amd/src/resource_catalog.js]: ✅ **Fixed in commit 0b8037b**

Added `JSON.parse` for roomids in `init()` and serialize as repeated parameters `roomids[]=1&roomids[]=2`.

File: amd/src/resource_catalog.js [hunk 5]
```diff
@@ -191,45 +193,53 @@ export default class extends BaseComponent {
...
+        if (!checkbox) {
+            window.console.warn(`Toggle checkbox not found for item ${element.id}`);
+            return;
+        }
+        checkbox.checked = element.active;
+
+        const label = document.querySelector(`label[for="resource-active-${element.id}"]`);
+        if (label) {
+            label.textContent = element.active ? 'Active' : 'Inactive';
+        }
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply amd/src/resource_catalog.js:244]: ✅ **Fixed in commit 85050e7**

Cache localized strings with `getString('active', 'core')` and `getString('inactive', 'core')` in `create()`.

File: amd/src/resource_catalog.js [hunk 6]
```diff
@@ -191,45 +193,53 @@ export default class extends BaseComponent {
...
+                amountCell.innerHTML = '<span class="badge badge-secondary">Unlimited</span>';
+            } else {
+                amountCell.innerHTML = `<span class="badge badge-info">${item.amount}x</span>`;
             }
+        } else if (fieldPart === 'name') {
+            const nameSpan = document.querySelector(`span[data-bookit-resource-tabledata-name-id="${item.id}"]`);
+            nameSpan.textContent = item.name;
+        } else if (fieldPart === 'description') {
+            const descSpan = document.querySelector(`small[data-bookit-resource-tabledata-description-id="${item.id}"]`);
+            descSpan.innerHTML = item.description || '';
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply amd/src/resource_catalog.js:291]: ✅ **Fixed in commit 8728643**

Added null checks for all DOM elements in `_replaceRenderedItem()` and handle description added/removed case.

File: amd/src/resource_catalog.js [hunk 7]
```diff
@@ -73,7 +74,8 @@ export default class extends BaseComponent {
                     description: itemEl.dataset.itemDescription || '',
                     categoryid: parseInt(itemEl.dataset.itemCategoryid),
                     amount: parseInt(itemEl.dataset.itemAmount) || 0,
-                    amountirrelevant: itemEl.dataset.itemAmountirrelevant === 'true',
+                    amountirrelevant: itemEl.dataset.itemAmountirrelevant === '1',
+                    active: itemEl.dataset.itemActive === '1',
                     sortorder: parseInt(itemEl.dataset.itemSortorder) || 0,
                 };
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply amd/src/resource_catalog.js:95]: ✅ **Already fixed by commit 0b8037b**

roomids parsing added in init(), description properly read from data attribute.

File: templates/resource_item_row.mustache [hunk 8]
```diff
@@ -85,12 +85,22 @@
...
-            <span class="badge badge-secondary">{{#str}}inactive, core{{/str}}</span>
-        {{/active}}
+        <div class="custom-control custom-switch">
+            <input type="checkbox"
+                   class="custom-control-input"
+                   data-action="toggle-active"
+                   data-item-id="{{id}}"
+                   {{#active}}checked{{/active}}
+                   id="resource-active-{{id}}">
+            <label class="custom-control-label" for="resource-active-{{id}}">
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply templates/resource_item_row.mustache:95]: ℹ️ **No changes needed**

Verified Moodle 4.5 uses Bootstrap 4.6.2. Current markup (`custom-control/custom-switch`) is correct for BS4.

File: amd/src/resource_category.js [hunk 9]
```diff
@@ -55,8 +55,8 @@ export default class ResourceCategory extends BaseComponent {
         this.itemComponents = new Map();
         this.parentElement = element;
 
-        // Render immediately.
-        this._render();
+        // Note: _render() must be called explicitly when creating new categories.
+        // When initializing from existing DOM, categoryElement is set externally.
     }
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply amd/src/resource_category.js:60]: ✅ **Fixed in commit 031fea4**

Made `_renderCategory` and `_handleCategoryCreated` async, now awaits `component._render()`.

**Review by aro-ssystems — COMMENTED** [review]
@copilot-pull-request-reviewer All 9 review comments have been addressed. Ready for re-review.

File: amd/src/resource_catalog.js [hunk 10]
```diff
@@ -67,14 +68,30 @@ export default class extends BaseComponent {
...
                     name: itemEl.dataset.itemName || '',
                     description: itemEl.dataset.itemDescription || '',
                     categoryid: parseInt(itemEl.dataset.itemCategoryid),
                     amount: parseInt(itemEl.dataset.itemAmount) || 0,
-                    amountirrelevant: itemEl.dataset.itemAmountirrelevant === 'true',
+                    amountirrelevant: itemEl.dataset.itemAmountirrelevant === '1',
+                    active: itemEl.dataset.itemActive === '1',
                     sortorder: parseInt(itemEl.dataset.itemSortorder) || 0,
+                    roomids: roomids,
                 };
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply amd/src/resource_catalog.js:95]: Not fixing - description field is optional and rarely used with HTML. Risk of data corruption is minimal in practice. The formatted HTML display is intentional for rich text support.

If this becomes an issue, we can revisit by adding data-item-description-raw attribute.

File: amd/src/resource_category.js [hunk 11]
```diff
@@ -141,25 +175,19 @@ export default class ResourceCategory extends BaseComponent {
...
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply amd/src/resource_category.js]: ✅ **Fixed in commit b5dd8fb**

Made _handleItemCreated() async and added await for _renderItems() call. This ensures proper error handling and prevents race conditions when multiple items are created rapidly.

File: amd/src/resource_catalog.js [hunk 12]
```diff
@@ -191,44 +223,73 @@ export default class extends BaseComponent {
...
+            } else if (item.description) {
+                // Description was added but <small> element doesn't exist - need to re-render row.
+                const row = document.querySelector(`#resource-item-row-${item.id}`);
+                if (row) {
+                    // Find category component and trigger re-render.
+                    const categoryComponent = Array.from(this.categoryComponents.values())
+                        .find(cat => cat.categoryData.id === item.categoryid);
+                    if (categoryComponent) {
+                        categoryComponent._renderItems();
+                    }
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply amd/src/resource_catalog.js]: ✅ **Fixed in commit 0ef433d**

Made _replaceRenderedItem() async and added await for categoryComponent._renderItems(). Refactored description update logic into separate _updateDescriptionField() method to reduce nesting depth. Added try/catch for proper error handling.

File: amd/src/resource_catalog.js [hunk 13]
```diff
@@ -455,6 +522,67 @@ export default class extends BaseComponent {
...
+        }])[0]
+        .then((response) => {
+            this.reactive.stateManager.processUpdates(JSON.parse(response.data));
+            return;
+        })
+        .catch((error) => {
+            // Revert checkbox on error.
+            checkbox.checked = !newActiveState;
+            window.console.error('Toggle active error:', error);
+        });
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply amd/src/resource_catalog.js]: ✅ **Fixed in commit a6e795c**

Converted promise chain to async/await pattern with try/catch. Now the async function properly awaits Ajax completion before returning, making it reliable for callers who await the function.

File: amd/src/resource_category.js [hunk 14]
```diff
@@ -97,8 +97,9 @@ export default class ResourceCategory extends BaseComponent {
 
     /**
      * Render items in this category (table view: append rows to tbody).
+     * Used when creating new items dynamically.
      */
-    _renderItems() {
+    async _renderItems() {
         if (!this.categoryElement) {
             return;
         }
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply amd/src/resource_category.js:105]: ✅ **Fixed in commits b5dd8fb and 3ca86de**

_handleItemCreated() was fixed in commit b5dd8fb (made async with await).
_render() was fixed in commit 3ca86de (added await for _renderItems()).

Both methods now properly await async operations, preventing unhandled rejections and race conditions.

File: amd/src/resource_catalog.js [hunk 15]
```diff
@@ -270,6 +331,12 @@ export default class extends BaseComponent {
         tableView.addEventListener('click', async(e) => {
             const target = e.target.closest('button[data-action]');
             if (!target) {
+                // Check for toggle switches.
+                const toggle = e.target.closest('[data-action="toggle-active"]');
+                if (toggle) {
+                    // Don't preventDefault - let checkbox toggle naturally.
+                    await this._handleToggleActive(toggle);
+                }
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply amd/src/resource_catalog.js]: ✅ **Fixed in commit db7eca2**

Changed event delegation from click to change event for toggle switches. The change event fires when checkbox value changes regardless of click target (checkbox or label), properly handling all toggle interactions.

File: templates/resource_item_row.mustache [hunk 16]
```diff
@@ -85,12 +85,22 @@
...
+                {{#active}}
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply templates/resource_item_row.mustache:96]: ✅ **Fixed via PR description update**

Updated PR description to clarify Bootstrap 4 is used (custom-control custom-switch markup). This is correct for Moodle 4.5 which uses Bootstrap 4.6.2.

File: amd/src/resource_catalog.js [hunk 17]
```diff
@@ -455,6 +522,67 @@ export default class extends BaseComponent {
...
+        const formDataString = params.toString();
+
+        // Submit via Ajax.
+        Ajax.call([{
+            methodname: 'core_form_dynamic_form',
+            args: {
+                formdata: formDataString,
+                form: 'mod_bookit\\form\\edit_resource_form'
+            }
+        }])[0]
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply amd/src/resource_catalog.js]: ✅ **Fixed via PR description update**

Updated PR description to match actual implementation. Changed from "ModalForm pattern" to "direct Ajax call" and explained the actual submission flow via Ajax.call() to core_form_dynamic_form.

File: amd/src/resource_catalog.js [hunk 18]
```diff
@@ -509,7 +636,12 @@ export default class extends BaseComponent {
      *
      * @param {Object} categoryData - Category data
      */
-    _renderCategory(categoryData) {
+    /**
+     * Render a new category.
+     *
+     * @param {Object} categoryData - Category data object
+     */
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply amd/src/resource_catalog.js:660]: ✅ **Fixed in commit 5191ed4**

Removed duplicate docblock. Kept the more descriptive one (Render a new category component) which is consistent with terminology used throughout the code.

File: amd/src/resource_catalog.js [hunk 19]
```diff
@@ -455,6 +546,65 @@ export default class extends BaseComponent {
...
+        const state = this.reactive.state;
+        const item = state.items.get(itemId);
+
+        if (!item) {
+            return;
+        }
+
+        // Get the new active state from the checkbox.
+        const newActiveState = checkbox.checked;
+
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply amd/src/resource_catalog.js:565]: minor, won't fix

File: amd/src/resource_catalog.js [hunk 20]
```diff
@@ -191,45 +223,96 @@ export default class extends BaseComponent {
...
+    /**
+     * Update description field in the DOM.
+     *
+     * @param {Object} item - Item data
+     */
+    async _updateDescriptionField(item) {
+        const descSpan = document.querySelector(`small[data-bookit-resource-tabledata-description-id="${item.id}"]`);
+        if (descSpan) {
+            descSpan.innerHTML = item.description || '';
```
[the remark this answers is not in the document]
Comment (aro-ssystems) [reply amd/src/resource_catalog.js:291]: minor, won't fix
